### PR TITLE
ntlm: support version 2 on 32-bit platforms and fix negotiated flags usage

### PR DESCRIPTION
**Pull request by monnerat**
This does not affect sspi builds.

I successfully performed as many tests as my non-M$ environment allows to, but this must be tested against a genuine M$ server, although I'm quite confident in it.

Thanks for considering it.

### CODE REVIEW

File: lib/vauth/ntlm.c [hunk 1]
```diff
@@ -591,21 +600,14 @@ CURLcode Curl_auth_create_ntlm_type3_message(struct Curl_easy *data,
...
-
-#define CURL_MD5_DIGEST_LENGTH 16 /* fixed size */
-
-  /* We don't support NTLM2 if we don't have USE_NTRESPONSES */
-  if(ntlm->flags & NTLMFLAG_NEGOTIATE_NTLM_KEY) {
+# else /* defined(USE_NTLM_V2) */
     unsigned char ntbuffer[0x18];
     unsigned char tmp[0x18];
-    unsigned char md5sum[CURL_MD5_DIGEST_LENGTH];
+    unsigned char md5sum[MD5_DIGEST_LENGTH];
```
Comment (bagder) [inline lib/vauth/ntlm.c:606]: Why this change? We provide our own define to make sure it exists under a single name. There's no guarantee there's a `MD5_DIGEST_LENGTH` define provided in systems! See d80d419d3e that introduced our name.

**Comment by bagder** [comment]
> this must be tested against a genuine M$ server

We need to figure out how to get someone to run such tests! I don't have any such servers/services to test against and I don't know who does! :thinking:

File: lib/vauth/ntlm.c (hunk 1 above)
Comment (monnerat) [reply lib/vauth/ntlm.c:606]: You're right, thanks for pointing it.
I intended to replace by `MD5_DIGEST_LEN` which is defined in `curl_md5.h`. Will fix it.

**Comment by monnerat** [comment]
> We need to figure out how to get someone to run such tests! I don't have any such servers/services to test against and I don't know who does!

We're both linux guys, that's why we are in the same situation.
If the initial ntlm developer is still around, maybe we should ask him/her for a review. Else a W$ contributor.

I would prefer not to have it committed for "testing by the community".
One thing is sure: current flag usage is wrong!

File: lib/vauth/ntlm.c (hunk 1 above)
Comment (monnerat) [reply lib/vauth/ntlm.c]: Fixed by using `MD5_DIGEST_LEN`.

**Comment by bagder** [comment]
> If the initial ntlm developer is still around, maybe we should ask him/her for a review. Else a W$ contributor.

I wrote the initial NTLM implementation in curl...

> One thing is sure: current flag usage is wrong!

... I believe you, but still it has been right enough to work pretty well for a lot of users over the years.

File: lib/vauth/ntlm.c (hunk 1 above)
Comment (bagder) [reply lib/vauth/ntlm.c]: Do you remove the `CURL_MD5_DIGEST_LENGTH` define as well then?

Comment (monnerat) [reply lib/vauth/ntlm.c]: Yes, because it was local to vauth/ntlm.c.

**Comment by monnerat** [comment]
> I wrote the initial NTLM implementation in curl...

Congratulations! I guess it was not that easy when no official documentation was released.
And how did you test it?

**Comment by bagder** [comment]
Other users at the time tested it against actual NTLM-using servers before I wrote up the test cases that verifiy that we maintain the same behavior.

**Comment by monnerat** [comment]
> Other users at the time tested it against actual NTLM-using servers before I wrote up the test cases that verifiy that we maintain the same behavior.

I had to adapt the tests as the NTLM response in type 3 message changes. I did it manually, checking that only the targeted message part was affected.

I guess NTLM is not much used nowadays.

**Comment by bagder** [comment]
> I guess NTLM is not much used nowadays.

According to our user survey 2020, 9.9% of our users use it, down just a little from 10.3% the year before. I think that's a pretty sizable share after all.

**Comment by monnerat** [comment]
> 9.9% of our users use it

Maybe we should call them for evaluation, via the mailing list for example ?

**Comment by MarcelRaad** [comment]
FWIW, I'm one of those who use NTLM (via SSPI though), but NTLMv2 only would suffice for me.

I can try to find a test server and test this patch.

**Comment by monnerat** [comment]
Many thanks Marcel,
In the meantime, I managed to setup a minimal IIS7 for testing.
The first tests resultswith it are not satisfactory, thus I think we should put it on hold for a few days.
I'll come back to your kind proposal when I feel it more mature.

**Comment by monnerat** [comment]
OK, I performed additional test with IIS7 and it works.

Here is a table of what is used depending on curl and server capabilities:

| curl \ server | V1ext | V1ext + V2 | V2 | V1 only |
| - | - | - | - | - |
| **V1ext** | V1ext | V1ext | V1 | V1 |
| **V1ext + V2** | V2 **!** | V2 | V1 | V1 |
| **V2** | V2 **!** | V2 | V1 | V1 |
| **V1 only** | V1 | V1 | V1 | V1 |

Entries marked with **!** are wrong and will fail, but these situations will probably never occur.
V2 cannot be negotiated: this explains oddities in this table.
By default curls supports V2 and V1ext (this can be changed by altering defines in `lib/curl_ntlm_core.h`).
Of course, this can fail if the security level used is not high enough for the server.
This should at least be compatible with M$ servers, samba and saslauthd.

It should be ready now.

**Comment by MarcelRaad** [comment]
The `V2` column looks strange. Shouldn't the `V1ext + V2` and `V2` rows contain `V2` instead of `V1`?

**Comment by monnerat** [comment]
> Shouldn't the V1ext + V2 and V2 rows contain V2 instead of V1?

No because as V2 cannot be negotiated, curl does not see server supports V2. We thus assume that V2 is supported only of V1ext is. IMHO this is the safest way.

**Comment by MarcelRaad** [comment]
OK, makes sense. I thought that V1ext + V2 looked the same to the client instead of V1 + V2.

**Comment by monnerat** [comment]
> I thought that V1ext + V2 looked the same to the client instead of V1 + V2.

No. In fact a V2-aware server does not care about the extended security flag when determining V2 is used: it only checks the NTLM response length.
The unknown here is: is the server V2-aware ?

The only other solution to the problem is to never use V2 (this will also remove a lot of code), but I don't think you will feel satisfied with it :-)

**Review by MarcelRaad — APPROVED** [review]

**Comment by bagder** [comment]
Thanks!

**Comment by monnerat** [comment]
Thanks for pulling.
